### PR TITLE
add kmeans++

### PR DESCRIPTION
**Pull request by ViperMiniQ**
Original paper: https://theory.stanford.edu/~sergei/papers/kMeansPP-soda.pdf
Original author presentation (with pictures and graphs): https://theory.stanford.edu/~sergei/slides/BATS-Means.pdf
Greedy kmeans++ (for the L value): https://drops.dagstuhl.de/storage/00lipics/lipics-vol173-esa2020/LIPIcs.ESA.2020.18/LIPIcs.ESA.2020.18.pdf

For reviewer(s), scikit implementation: https://github.com/scikit-learn/scikit-learn/blob/main/sklearn/cluster/_kmeans.py#L174
(keep in mind if you want to test scikit results, the idea with kmeans++ is that it randomly selects initial centers and thus you can get different results across runs, but to achieve that with scikit KMeans, you have to pass it rand generator – if not, it will use one, but with a fixed seed value, so you will get the same result across multiple runs) 

Kmeans++ differs only in the initial center selection, the rest is the same. On average, it converges faster (requires less iterations) than the farthest points initialization.

I could not come up with a test case with 3 or more clusters in reasonable time (not execution, but thinking problem ;)), so I didn’t include any. I guess that’s the reason scikit implementation can be passed seed value for the generator as with it the algorithm becomes deterministic if we know all the values and the result can be calculated.

### CODE REVIEW

File: src/analysis/processing/qgsalgorithmkmeansclustering.cpp [hunk 1]
```diff
@@ -298,6 +317,140 @@ void QgsKMeansClusteringAlgorithm::initClusters( std::vector<Feature> &points, s
...
+    distances[i] = distance;
+    totalError += distance;
+    if ( qgsDoubleNear( distance, 0 ) )
+    {
+      duplicateCount++;
+    }
+  }
+  if ( feedback && duplicateCount > 1 )
+  {
+    feedback->pushInfo( QObject::tr( "There are at least %n duplicate input(s), the number of output clusters may be less than was requested", nullptr, duplicateCount ) );
```
Comment (nyalldawson) [inline src/analysis/processing/qgsalgorithmkmeansclustering.cpp:357]:
```suggestion
    feedback->pushWarning( QObject::tr( "There are at least %n duplicate input(s), the number of output clusters may be less than was requested", nullptr, duplicateCount ) );
```

**Comment by nyalldawson** [comment]
@ViperMiniQ can you rebase and address the failing check?

**Review by nyalldawson — APPROVED** [review]

**Comment by ViperMiniQ** [comment]
clang tidy error: file '/root/QGIS/src/core/qgis.h' has been modified since the precompiled header '/root/QGIS/build/src/analysis/CMakeFiles/qgis_analysis.dir/cmake_pch.hxx.pch' was built: mtime changed (was 1747381379, now 1747384863) [clang-diagnostic-error]

Not sure what I can do here, seems to be build problem.
